File: helix/src/main/java/glitch/helix/service/request/UserFollowRequest.java
package glitch.helix.service.request;

import glitch.api.http.HttpClient;
import glitch.api.http.Routes;
import glitch.helix.object.json.Follow;
import glitch.helix.object.json.Follows;
import glitch.service.AbstractRequest;
import javax.annotation.Nullable;
import reactor.core.publisher.Mono;

public class UserFollowRequest extends AbstractRequest<Follow, Follows> {
    @Nullable
    private final Long fromId;
    @Nullable
    private final Long toId;
    private String after;
    private Integer first;

    public UserFollowRequest(HttpClient httpClient, Long fromId, Long toId) {
        super(httpClient, Routes.get("/users/follows").newRequest());
        this.fromId = fromId;
        this.toId = toId;
    }

    public UserFollowRequest setAfter(String after) {
        this.after = after;
        return this;
    }

    public UserFollowRequest setFirst(Integer first) {
        this.first = first;
        return this;
    }

    @Override
    public Mono<Follows> get() {
        if (after != null) {
            request.queryParam("after", after);
        }

        if (first != null && first > 0 && first <= 100) {
            request.queryParam("first", first);
        }

        if (fromId != null) {
            request.queryParam("from_id", fromId);
        }

        if (toId != null) {
            request.queryParam("to_id", toId);
        }

        return httpClient.exchangeAs(request, Follows.class);
    }
}

File: helix/src/main/kotlin/io/glitchlib/api/service/request/UserFollowRequest.kt
package io.glitchlib.api.service.request

import io.glitchlib.internal.http.AbstractRequest
import io.glitchlib.internal.http.HttpRequest

class UserFollowRequest internal constructor() : AbstractRequest {
    private var limit: Int? = null
    private var after: String? = null
    private var before: String? = null

    private var from: Long? = null
    private var to: Long? = null

    fun limit(limit: Int) = apply {
        this.limit = limit
    }

    fun after(after: String) = apply {
        this.after = after
    }

    fun before(before: String) = apply {
        this.before = before
    }

    fun from(from: Long) = apply {
        this.from = from
    }

    fun to(to: Long) = apply {
        this.to = to
    }

    override fun invoke(): HttpRequest.() -> Unit = {
        if (limit != null && limit!! in 1..100) addQueryParameters("first", limit!!.toString())
        if (after != null) addQueryParameters("after", after!!)
        if (before != null) addQueryParameters("before", before!!)
        if (from != null) addQueryParameters("from_id", from.toString())
        if (to != null) addQueryParameters("to_id", to.toString())
    }
}
